JavaIntegration/src/main/java/org/openzen/zencode/java/module/converters/JavaAnnotatedType.java

```diff
@@ -0,0 +1,171 @@
+package org.openzen.zencode.java.module.converters;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.*;
+import java.util.Arrays;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public final class JavaAnnotatedType implements AnnotatedElement, Type {
+	public enum ElementType {
+		ANNOTATED_PARAMETERIZED_TYPE,
+		ANNOTATED_TYPE,
+		CLASS,
+		GENERIC_ARRAY,
+		PARAMETERIZED_TYPE,
+		TYPE_VARIABLE,
+		WILDCARD
+	}
+
+	private final ElementType elementType;
+	private final AnnotatedElement annotatedElement;
+	private final Type type;
+
+	private JavaAnnotatedType(final ElementType elementType, final AnnotatedElement element, final Type type) {
+		this.elementType = elementType;
+		this.annotatedElement = element;
+		this.type = type;
+	}
+
+	public static JavaAnnotatedType of(final Object element) {
+		if (element instanceof Type && element instanceof AnnotatedElement) {
+			return or(ofBoth(element), invalid(element));
+		}
+		if (element instanceof Type) {
+			return or(of((Type) element), invalid(element));
+		}
+		if (element instanceof AnnotatedElement) {
+			return or(of((AnnotatedElement) element), invalid(element));
+		}
+		throw invalid(element);
+	}
+
+	public static JavaAnnotatedType[] of(final Object[] element) {
+		return Arrays.stream(element).map(JavaAnnotatedType::of).toArray(JavaAnnotatedType[]::new);
+	}
+
+	private static JavaAnnotatedType or(final JavaAnnotatedType result, final RuntimeException ifNull) {
+		if (result == null) throw ifNull;
+		return result;
+	}
+
+	private static JavaAnnotatedType ofBoth(final Object element) {
```

Comment:
Maybe to make the expected type explicit (and thereby type-safer) you could do something like
```suggestion
	private static <T extends Type & AnnotatedElement> JavaAnnotatedType of(final T element) {
```